Source/items.cpp

```diff
@@ -4140,98 +4124,87 @@ void UseItem(int p, item_misc_id Mid, spell_id spl)
 
 bool StoreStatOk(ItemStruct *h)
 {
-	bool sf;
-
-	sf = true;
 	if (plr[myplr]._pStrength < h->_iMinStr)
-		sf = false;
+		return false;
 	if (plr[myplr]._pMagic < h->_iMinMag)
-		sf = false;
+		return false;
 	if (plr[myplr]._pDexterity < h->_iMinDex)
-		sf = false;
+		return false;
 
-	return sf;
+	return true;
 }
 
 bool SmithItemOk(int i)
 {
-	bool rv;
-
-	rv = true;
 	if (AllItemsList[i].itype == ITYPE_MISC)
-		rv = false;
+		return false;
 	if (AllItemsList[i].itype == ITYPE_GOLD)
-		rv = false;
+		return false;
 	if (AllItemsList[i].itype == ITYPE_STAFF && (!gbIsHellfire || AllItemsList[i].iSpell))
-		rv = false;
+		return false;
 	if (AllItemsList[i].itype == ITYPE_RING)
-		rv = false;
+		return false;
 	if (AllItemsList[i].itype == ITYPE_AMULET)
-		rv = false;
+		return false;
 
-	return rv;
+	return true;
 }
 
-int RndSmithItem(int lvl)
+template <bool (*Ok)(int), bool ConsiderDropRate = false>
+int RndVendorItem(int minlvl, int maxlvl)
 {
-	int i, ri;
 	int ril[512];
 
-	ri = 0;
-	for (i = 1; AllItemsList[i].iLoc != ILOC_INVALID; i++) {
+	int ri = 0;
+	for (int i = 1; AllItemsList[i].iLoc != ILOC_INVALID; i++) {
 		if (!IsItemAvailable(i))
 			continue;
+		if (AllItemsList[i].iRnd == IDROP_NEVER)
+			continue;
+		if (!Ok(i))
+			continue;
+		if (AllItemsList[i].iMinMLvl < minlvl || AllItemsList[i].iMinMLvl > maxlvl)

```

Comment:
This seams a bit backwards. Items define what range they work with and evaluate against the given parameters. But here you are comparing `iMinMlvl` to a range as if it was the items level. It also looks like `minlvl` is always `0`, so this would probably make more sens as:
```suggestion
		if (AllItemsList[i].iMinMLvl > lvl)
```

Reply:
SpawnPremium cares about the minimum level. Adria etc. don't.

Reply:
Ok, so this is in preparation of further cleanups?

Reply:
I've already done the cleanup: SpawnPremium also uses RndVendorItem now.

Reply:
Arh, missed that detail :)